Ansible: Fix connection plugins and add support for NETCONF

mweinelt opened the pull request:
<!-- Nixpkgs has a lot of new incoming Pull Requests, but not enough people to review this constant stream. Even if you aren't a committer, we would appreciate reviews of other PRs, especially simple ones like package updates. Just testing the relevant package/service and leaving a comment saying what you tested, how you tested it and whether it worked would be great. List of open PRs: <https://github.com/NixOS/nixpkgs/pulls>, for more about reviewing contributions: <https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions>. Reviewing isn't mandatory, but it would help out a lot and reduce the average time-to-merge for all of us. Thanks a lot if you do! -->
###### Motivation for this change


###### Things done

1. Patches Ansible to use the unwrapped version of `ansible-connection`, as it executes it through a python interpreter, which currently results in a SyntaxError. Fixes #71342 

2. Includes `ncclient` to add support for netconf connections to configure network hardware.

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Determined the impact on package closure size (by running `nix path-info -S` before and after)
  - Before: `/nix/store/8hz5c81n1cfgwczn267y1prdizhw55q8-python3.7-ansible-2.8.4	 220.9M`
  - After: `/nix/store/azxh8l7bvkga12yk10zm0qscc5ai6cah-python3.7-ansible-2.8.4	 228.4M`
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

###### Notify maintainers

cc @joamaki @costrouc

mweinelt committed:
ansible: Fix use of connection plugins

Ansible tries to execute the ansible-connection script as a python
script from within it's task executor. This does not work when it is
encapsulated in a shell script. Therefore remove the call to the python
interpreter and use the wrapped version directly.

Fixes #71342

mweinelt committed:
ansible: Add ncclient to support NETCONF connections

mweinelt commented:
Also https://github.com/NixOS/nixpkgs/pull/72370/files#diff-7fba719869598b74dcfbab4a83d750a5R32 looks like a no-op to me.

No `/usr` in https://github.com/ansible/ansible/blob/stable-2.8/lib/ansible/constants.py between v2.4..v2.9.

FRidh commented on pkgs/development/python-modules/ansible/default.nix:
```diff
@@ -29,6 +30,10 @@ buildPythonPackage rec {
+
+    # Use the unwrapped version of ansible-connection as ansible executes
+    # it directly through the python interpreter.
+    sed -i "s,ansible-connection,.ansible-connection-wrapped," lib/ansible/executor/task_executor.py
```
Removing `python` from
```
[python, ansible_connection, to_text(os.getppid())]
```
is likely more robust. It would also be worth upstreaming that change IMO.

mweinelt replied:
This would remove the guarantee that `ansible_connection` gets executed with the same Python version, wouldn't it?

It was like this before until https://github.com/ansible/ansible/commit/0feea669880e4387e7d243812fb8ba9e9d1bbb16 which was necessary to fix https://github.com/ansible/ansible/issues/35902, so I think they'd be reluctant to accept that change.

FRidh replied:
It is indeed not guaranteed that it is executed with the exact same interpreter. That is a lousy requirement anyway, because having the same interpreter does not guarantee that it will have the required packages, which is what their intent is. Instead they should rely on their executables already pointing to the right Python.

mweinelt replied:
Is this still an issue for you?

mweinelt committed:
ansible: Some cleanups

- boto was dropped from propagatedBuildInputs in 4af94d0f
- lib/ansible/constants.py does not have "/usr" anywhere in v2.4..v2.9
